### PR TITLE
[Lock] Refine the contract and implementation for Stores to handle Not Expirable Store cases.

### PR DESCRIPTION
**Pull request by GaneshChandrasekaran-zz**
The Stores contract and implementation doesn't clearly distinguish between stores that support expirable stores and ones which don't. This change attempts to improve that for better usage of the functions by the services without running into runtime issues.

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes 
| Fixed tickets | #28696 
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by GaneshChandrasekaran** [comment]
@jderusse - this is now ready for review.

Also, I can figure out why it's failing in this environment
https://travis-ci.org/symfony/symfony/jobs/437228068

I have a feeling that it's reading files from `vendor` directory but I am unable to confirm that from the stack trace.

File: src/Symfony/Component/Lock/Store/CombinedStore.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php]: Should be take into account in`$this->strategy->isMet($successCount, $storesCount)`

File: src/Symfony/Component/Console/Command/LockableTrait.php
Comment (jderusse) [inline src/Symfony/Component/Console/Command/LockableTrait.php]: Related to this PR?

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Should be catch in `Lock::refresh`

Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Are you sure? Because in the comment linked below I was suggesting it should be bubbled up in `Lock::refresh` as well because the client should not create a lock with `ttl`

https://github.com/symfony/symfony/issues/28696#issuecomment-427000248

File: src/Symfony/Component/Console/Command/LockableTrait.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Console/Command/LockableTrait.php]: Yes, it is. I didn't notice this until the build started failing and showed errors in `Console` component. With this change, when creating a lock using `Factory` we should override the `ttl` to be `null` for stores that don't support expiration because they throw the new exception.

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Okay - so `refresh` can be called outside of `save` as well since a lock can be refreshed at anytime. Yes, in that case I should catch and do nothing.

Comment (jderusse) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: yes, you're right. 

Issue is, services are not able to guess if the StoreInterface is able to extends  the TTL until calling the method and triggering an exception..

For instance, I could have a code running Semaphore in test/dev environement and redis in production and always calling `Lock->refesh()`...

I wonder if dealing with 2 interfaces wouldn't be easiest and simplest than this 2 exceptions...

```
interface ExpirableStoreInterface extends StoreInterface {
    public function putOffExpiration(Key $key, $ttl);
    public function saveIn(Key $key, $ttl);
}
class Lock {
  // either adding a new isRefreshable() or ignoring exception from not expirable stores...
  public function isRefreshable(): bool {return $this->store instanceOf ExpirableStoreInterface;}
}
```

In that case we have to duplicate/extends decoratorStore (CombinedStore and RetryTillSaveStore) in order to expose the right interface to the user
``` 
class CombinedStore implements StoreInterface {}
class ExpirableCombinedStore extends CombinedStore implements ExpirableStoreInterface {}

class RetryTillSaveStore implements StoreInterface {}
class ExpirableRetryTillSaveStore extends RetryTillSaveStore implements ExpirableStoreInterface {}
```
Or we can add a method `supportsExpiration` in that next interace. But it's weird
```
class ExpirableRetryTillSaveStore implements ExpirableStoreInterface {
  public function supportsExpiration(): bool {
    return $this->decorated instanceOf ExpirableStoreInterface && $this->decorated->supportsExpiration()
  }
}
```

File: src/Symfony/Component/Lock/Store/CombinedStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/CombinedStore.php]: I have now removed throwing exceptions and added new `ExpirableStoreInterface` instead so reverting this change.

File: src/Symfony/Component/Console/Command/LockableTrait.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Console/Command/LockableTrait.php]: Reverted. See above.

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Your issue makes absolute sense. I also think it makes sense have a cleaner contract depending on what type of functionality the stores can support. The issue with leaving a noop in `putOffExpiration` will definitely cause the same issue you mentioned where if a user calls a `ttl` with `null` then an `InvalidArgumentException` is thrown.

Adding the new methods you suggested `isRefreshable` and the new `ExpirableStoreInterface` definitely makes the implementation more clean. I have updated the diff with the new design and remove the `NotExpirableStoreException` as it doesn't seem to be required anymore. 

I can't see any other obvious issues with this other than the fact that for `CombinedStores` we cannot combine an `Expirable Store` and `Non-expirable Store`. But I also don't see a good use case of having those two together in the decorator. 

Let me know your thoughts on this design.

File: src/Symfony/Component/Lock/Store/ExpirableCombinedStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/ExpirableCombinedStore.php]: Yeah that’s what I was thinking too. I guess we should just go ahead ignore exception or do what other stores do. I am out for a week but can change that once I come back.

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php
Comment (OskarStark) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: why do you switch to `protected` here?

File: src/Symfony/Component/Lock/Tests/Store/ExpirableCombinedStoreTest.php
Comment (OskarStark) [inline src/Symfony/Component/Lock/Tests/Store/ExpirableCombinedStoreTest.php]: you could use `createMock()` here IIRC

File: src/Symfony/Component/Lock/Tests/Store/RetryTillSaveStoreTest.php
Comment (OskarStark) [inline src/Symfony/Component/Lock/Tests/Store/RetryTillSaveStoreTest.php]: s/$zookeeper_server/$zookeeperServer/

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Based on https://github.com/symfony/symfony/pull/28727#discussion_r223501776
We are just going to catch and ignore the new created exception `NotExpirableStoreException`.

File: src/Symfony/Component/Lock/Store/ExpirableCombinedStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/ExpirableCombinedStore.php]: Fixed.

File: src/Symfony/Component/Lock/Store/SemaphoreStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/SemaphoreStore.php]: Pinging @jderusse for review :)

File: src/Symfony/Component/Lock/Store/CombinedStore.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/CombinedStore.php]: You should either increment `$successCount` or create a new `$ignoreCount` variable in order to define whether or not the strategy is met.

File: src/Symfony/Component/Lock/Store/FlockStore.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/FlockStore.php]: I wonder if it's should be treated like a BC break. ping @nicolas-grekas

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: IMHO we shouldn't catch the exception here.

**Review by jderusse — APPROVED** [review]

File: src/Symfony/Component/Lock/Store/RetryTillSaveStore.php
Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: I was thinking of the case where you mentioned if in DEV the application is running an `expirable` store and in production if it's changed to `non-expirable` store then this should deal with that.

Comment (jderusse) [reply src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: My initial point was that "a client with a Lock using a generic StoreInterface is not able to know whether the store will throw an exception or not". But now things changed: all stores  (that does not support the feature) are throwing an exception and that exception is catch by the `Lock` class. This decorator could throw an exception here

Comment (GaneshChandrasekaran) [reply src/Symfony/Component/Lock/Store/RetryTillSaveStore.php]: Ahh yes that makes sense. Fixed.

File: src/Symfony/Component/Lock/Store/FlockStore.php
Comment (fabpot) [reply src/Symfony/Component/Lock/Store/FlockStore.php]: Looks like a BC break to me, but might be worth doing anyway instead of silently ignoring the call.

File: src/Symfony/Component/Lock/Tests/Store/CombinedStoreTest.php
Comment (fabpot) [inline src/Symfony/Component/Lock/Tests/Store/CombinedStoreTest.php]: Small typo, should be `$zookeeperServer`

**Review by fabpot — APPROVED** [review]
Looks good to me with two small comments. @jderusse I think we need your approval before merging.

File: src/Symfony/Component/Lock/Lock.php
Comment (jderusse) [inline src/Symfony/Component/Lock/Lock.php]: The method `refresh` is called internally when calling `acquire` with a non-null TTL (which is the default case when using the Factory with default arguments)

When using a NotExpirableStore (like flock) the logger will be full of notice.
Same comment when combining a non-expirable store and expirable one (flock + redis for instance)

IMHO You should either silent this exception when calling `acquire` or update the factory to set a null TTL when the store does not support expiration.

Would be great from a user point of view to known whether or not the store supports expiration.
```
interface StoreInterface {
  public function save(Key $key);
  public function waitAndSave(Key $key);
  public function delete(Key $key);
  public function exists(Key $key);
}

interface ExpirableStoreInterface extends StoreInterface {
  public function supportsExpiration(): bool
  public function putOffExpiration(Key $key, $ttl);
}
```

**Comment by jderusse** [comment]
closed by https://github.com/symfony/symfony/pull/32299 @fabpot

**Pull request closed without merging**
